### PR TITLE
[ML] Compute retrained tree weight shrinkage by cross-validation

### PR DESCRIPTION
**Pull request by tveasey**
The best shrinkage to use for trees we retrain is not expected to match that for train. The incremental loss function already builds in a mechanism for keeping the weights small by penalising the difference in predictions from the old tree. Furthermore, since we retrain fewer trees than we train it is likely that we need less shrinkage in order to impact the new training data loss sufficiently. Testing on both target drift and out of training domain scenarios shows that computing the best shrinkage by cross-validation and searching a wide range up to and including no shrinkage gives *greatly* improved QoR. This also adds new unit tests for out of training domain data and incremental binomial logistic regression.

### CODE REVIEW

File: lib/maths/CBoostedTreeImpl.cc [hunk 1]
```diff
@@ -978,6 +977,10 @@ CBoostedTreeImpl::updateForest(core::CDataFrame& frame,
         const auto& treeToRetrain = m_BestForest[index];
         const auto& treeWhichWasRetrained = retrainedTrees.back();
 
+        double eta{index < m_BestForest.size()
```
Comment (tveasey) [inline lib/maths/CBoostedTreeImpl.cc:980]: I want to anticipate the case that we consider adding trees as well. Although currently we only ever use `m_RetrainedTreeEta`.

**Comment by valeriy42** [comment]
> Testing on both target drift and out of training domain scenarios shows that computing the best shrinkage by cross-validation and searching a wide range up to and including no shrinkage gives greatly improved QoR.
:rocket: 

It would be great if you could show some of the QoR improvements as a comment to this PR.

File: include/maths/CBoostedTreeFactory.h [hunk 2]
```diff
@@ -249,25 +251,33 @@ class MATHS_EXPORT CBoostedTreeFactory final {
...
-    //! Set the initial values for the various hyperparameters.
+    //! Set the initial values for hyperparameters.
     void initializeHyperparameters(core::CDataFrame& frame);
 
-    //! Setup before initializing unset hyperparameters.
+    //! Setup before setting initial values for hyperparameters.
     void initializeHyperparametersSetup(core::CDataFrame& frame);
 
-    //! Estimate a good search bounding box regularisation hyperparameters.
+    //! Estimate a good initial value and search bounding box for regularisation
```
Comment (valeriy42) [inline include/maths/CBoostedTreeFactory.h:260]:
```suggestion
    //! Estimate a good initial value and bounding box to search for regularisation
```

**Review by valeriy42 — APPROVED** [review]
LGTM. Good work on the new unit tests!

**Comment by tveasey** [comment]
__Extrapolation__

These are derived from the unit test `CBoostedTreeTest/testMseIncrementalForOutOfDomain` where I generate new training data entirely outside the training data domain. The results are averaged over 5 different random samples each for piecewise constant, linear and non-linear functions. The following show test errors for old and new data after incremental training before and after this change.

| Data set | Before average MSE old data | Before average MSE new data | After average MSE old data | After average MSE new data |
| --- | --- | --- | --- | --- |
| Piecewise constant | 24 | 30 | 23 | 23 |
| Linear | 92 | 372 | 87 | 119 |
| Non-linear | 116 | 859 | 36 | 158 |

**Comment by tveasey** [comment]
__Drift__

These are derived from the unit test `CBoostedTreeTest/testMseIncrementalForTargetDrift` where I generate new training data inside the training data domain from a perturbed target function. The results are averaged over 5 different random samples each for piecewise constant, linear and non-linear functions. The following show test errors for old and new data after incremental training before and after this change.

| Data set | Before average MSE old data | Before average MSE new data | After average MSE old data | After average MSE new data |
| --- | --- | --- | --- | --- |
| Piecewise constant | 8 | 28 | 9 | 22 |
| Linear | 15 | 41 | 16 | 31 |
| Non-linear | 41 | 465 | 58 | 269 |

Note in this case there is nothing we can do to avoid increasing old prediction errors somewhat because there is a discrepancy. This shows the need to manage the relative importance of old and new training data. This is really a question for the user and should be encoded in weights for each training example. That change is left for a follow on PR.

File: include/maths/CBoostedTreeFactory.h (hunk 2 above)
Comment (tveasey) [reply include/maths/CBoostedTreeFactory.h]: [94d1399](https://github.com/elastic/ml-cpp/pull/1943/commits/94d13992d916a3137bb84530cc3c72b322255b74)

**Comment by tveasey** [comment]
Test failure is a timing issue and unrelated to this PR.
